src/main/java/edu/kit/compiler/codegen/pattern/OperandPattern.java

```diff
@@ -66,34 +121,208 @@
+            var immediateMatch = IMMEDIATE.match(node, matcher);
+            if (immediateMatch.matches()) {
+                var immediate = immediateMatch.getOperand();
+                var operand = Operand.immediateRegister(immediate,
+                        matcher.getNewRegister(immediate.getSize()));
+                return OperandMatch.some(operand, Collections.emptyList());
             } else {
-                return OperandMatch.none();
+                var register = matcher.getRegister(node);
+                if (register.isPresent()) {
+                    var predecessors = List.of(node);
+                    var operand = Operand.register(node.getMode(), register.get());
+                    return OperandMatch.some(operand, predecessors);
+                } else {
+                    return OperandMatch.none();
+                }
             }
         }
     }
 
     @NoArgsConstructor(access = AccessLevel.PRIVATE)
     public static final class MemoryPattern implements Pattern<OperandMatch<Memory>> {
+
+        private static final Pattern<OperandMatch<Immediate>> OFFSET = OperandPattern.immediate(RegisterSize.DOUBLE);
+        private static final Pattern<OperandMatch<Immediate>> SCALE = OperandPattern.addressScale();
+        private static final Pattern<OperandMatch<Register>> REGISTER = OperandPattern.register();
+
         @Override
         public OperandMatch<Memory> match(Node node, MatcherState matcher) {
-            var register = matcher.getRegister(node);
-            if (node.getMode().equals(Mode.getP()) && register.isPresent()) {
-                var operand = Operand.register(node.getMode(), register.get());
-                return OperandMatch.some(Operand.memory(operand), List.of(node));
+            // We make some simplifying assumptions here. If these assumptions
+            // are not upheld, worse addressing modes may be used as a result.
+            // If arithmetic identities have been eliminated, these assumptions
+            // will always be upheld.
+            // - Constants should always be right side of addition (x + c)
+            // - Constants should always be right side of multiplication (x * c)
+            // - Subtractions of constants should not be used (x - c == x + (-c))
+
+            var nodes = AddressNodes.of(node);
+
+            var indexRight = matchMul(nodes.offset, nodes.firstRegister,
+                    nodes.secondRegister, matcher);
+            if (indexRight.matches()) {
+                return indexRight;
+            }
+
+            var indexLeft = matchMul(nodes.offset, nodes.secondRegister,
+                    nodes.firstRegister, matcher);
+            if (indexLeft.matches()) {
+                return indexLeft;
+            }
+
+            // todo (2,4,8) + 1 * %rax = (%rax,%rax,8)
+            var match = getMatch(nodes.offset, nodes.firstRegister,
+                    nodes.secondRegister, matcher);
+            if (match == null) {
```

Comment:
I might be mistaken, but to me it doesn't seem possible that `getMatch` returns `null`. Maybe this should rather be `match.matches()`?